crates/bevy_asset/src/handle.rs

```diff
@@ -1,472 +1,383 @@
-    Id(Uuid, u64),
-
-    /// A handle id of a pending asset.
-    AssetPathId(AssetPathId),
+/// Provides [`Handle`] and [`UntypedHandle`] _for a specific asset type_.
+/// This should _only_ be used for one specific asset type.
+#[derive(Clone)]
+pub struct AssetHandleProvider {
+    pub(crate) allocator: Arc<AssetIndexAllocator>,
+    pub(crate) drop_sender: Sender<DropEvent>,
+    pub(crate) drop_receiver: Receiver<DropEvent>,
+    pub(crate) type_id: TypeId,
 }
 
-impl From<AssetPathId> for HandleId {
-    fn from(value: AssetPathId) -> Self {
-        HandleId::AssetPathId(value)
-    }
+pub(crate) struct DropEvent {
+    pub(crate) id: InternalAssetId,
+    pub(crate) asset_server_managed: bool,
 }
 
-impl<'a> From<AssetPath<'a>> for HandleId {
-    fn from(value: AssetPath<'a>) -> Self {
-        HandleId::AssetPathId(AssetPathId::from(value))
+impl AssetHandleProvider {
+    pub(crate) fn new(type_id: TypeId, allocator: Arc<AssetIndexAllocator>) -> Self {
+        let (drop_sender, drop_receiver) = crossbeam_channel::unbounded();
+        Self {
+            type_id,
+            allocator,
+            drop_sender,
+            drop_receiver,
+        }
     }
-}
 
-impl<'a, 'b> From<&'a AssetPath<'b>> for HandleId {
-    fn from(value: &'a AssetPath<'b>) -> Self {
-        HandleId::AssetPathId(AssetPathId::from(value))
+    /// Reserves a new strong [`UntypedHandle`] (with a new [`UntypedAssetId`]). The stored [`Asset`] [`TypeId`] in the
+    /// [`UntypedHandle`] will match the [`Asset`] [`TypeId`] assigned to this [`AssetHandleProvider`].
+    pub fn reserve_handle(&self) -> UntypedHandle {
+        let index = self.allocator.reserve();
+        UntypedHandle::Strong(self.get_handle(InternalAssetId::Index(index), false, None, None))
+    }
+
+    pub(crate) fn get_handle(
+        &self,
+        id: InternalAssetId,
+        asset_server_managed: bool,
+        path: Option<AssetPath<'static>>,
+        meta_transform: Option<MetaTransform>,
+    ) -> Arc<InternalAssetHandle> {
+        Arc::new(InternalAssetHandle {
+            id: id.untyped(self.type_id),
+            drop_sender: self.drop_sender.clone(),
+            meta_transform,
+            path,
+            asset_server_managed,
+        })
+    }
+
+    pub(crate) fn reserve_handle_internal(
+        &self,
+        asset_server_managed: bool,
+        path: Option<AssetPath<'static>>,
+        meta_transform: Option<MetaTransform>,
+    ) -> Arc<InternalAssetHandle> {
+        let index = self.allocator.reserve();
+        self.get_handle(
+            InternalAssetId::Index(index),
+            asset_server_managed,
+            path,
+            meta_transform,
+        )
+    }
+}
+
+/// The internal "strong" [`Asset`] handle storage for [`Handle::Strong`] and [`UntypedHandle::Strong`]. When this is dropped,
+/// the [`Asset`] will be freed. It also stores some asset metadata for easy access from handles.
+#[derive(TypePath)]
+pub struct InternalAssetHandle {
```

Comment:
Would a typed version of this make sense?

Reply:
I can see the motivation (remove the stored TypeId), but I don't think its the right call. We need the StrongHandle to represent the lifetime of the asset. But "owners" of this handle can be both typed and untyped (including the AssetServer itself, which is untyped). I don't see a way to make this "typed" without factoring out an "untyped" version and hoisting dynamic vs static type information to a wrapper type (and copying that around for each handle, increasing the cost of handle clones and the size of handles on the stack).